Pass basic auth to all system calls

burtonr opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This changeset enables passing the basic auth credentials
to all /system/ calls to allow upstream providers to
perform authorization checks independent of the gateway.

This is essential for some providers, like Swarm, where
the system is accessible on the same network, and not
protected via the gateway

Signed-off-by: Burton Rheutan <rheutan7@gmail.com>

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [X] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
Resolves #836 

This will enable the work on faas-swarm https://github.com/openfaas/faas-swarm/issues/31

This will allow providers to perform authorization checks when being called directly for additional security. 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
I've run this locally running Swarm to verify that the additional headers are being passed, but without the other services using this header, it is a non-breaking change.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

alexellis commented on gateway/server.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) Alex Ellis 2017. All rights reserved.
```
Why have you removed my personal copyright from this file?

burtonr replied:
I only changed it to the broader "OpenFaaS Project" to match other files as I thought there weren't supposed to be individual names in the copyright. I will add your name back, in addition to the overarching "OpenFaaS Author(s)" line

alexellis commented on gateway/server.go:
```diff
@@ -117,6 +119,12 @@ func main() {
 			handlers.DecorateWithBasicAuth(faasHandlers.ListFunctions, credentials)
 		faasHandlers.ScaleFunction =
 			handlers.DecorateWithBasicAuth(faasHandlers.ScaleFunction, credentials)
+		faasHandlers.Alert =
```
If I merge this change then the auto-scaling will fail because Prometheus does not send auth strings to the scaling endpoint.

alexellis commented on gateway/server.go:
```diff
@@ -117,6 +119,12 @@ func main() {
 		faasHandlers.ScaleFunction =
 			handlers.DecorateWithBasicAuth(faasHandlers.ScaleFunction, credentials)
+		faasHandlers.Alert =
+			handlers.DecorateWithBasicAuth(faasHandlers.Alert, credentials)
```
I think this would break since I'm not sure the queue worker sends auth

alexellis commented on gateway/server.go:
```diff
@@ -117,6 +119,12 @@ func main() {
+		faasHandlers.AsyncReport =
+			handlers.DecorateWithBasicAuth(faasHandlers.AsyncReport, credentials)
+		infoHandler = handlers.DecorateWithBasicAuth(infoHandler, credentials)
+		queryFunction = handlers.DecorateWithBasicAuth(queryFunction, credentials)
```
Did this not have auth enabled previously?

burtonr commented:
@alexellis I've updated to exclude the `Alert` and `AsyncReport` handlers from the auth check. Also, fixed the copyright line to retain your name (as was originally) and added "OpenFaaS Author(s)"

In doing so, I first verified that, yes those changes caused the scaling to no longer function. I've verified that the scaling _does_ function as expected now. 

The only 2 calls that are added to the auth check are the info and the query.
